crates/sui-indexer-restorer/src/snapshot.rs

```diff
@@ -0,0 +1,214 @@
+            ..Default::default()
+        };
+
+        let m = MultiProgress::new();
+        let snapshot_reader = StateSnapshotReaderV1::new(
+            args.start_epoch,
+            &remote_store_config,
+            &local_store_config,
+            usize::MAX,
+            NonZeroUsize::new(args.concurrency).unwrap(),
+            m,
+            true, // skip_reset_local_store
+        )
+        .await?;
+
+        let database_url = Url::parse(&args.database_url)?;
+        let db_args = DbArgs::with_url(database_url);
+        let db = Db::new(db_args).await?;
+
+        Ok(Self {
+            restore_args: args.clone(),
+            snapshot_reader,
+            db,
+            next_checkpoint_after_epoch,
+        })
+    }
+
+    pub async fn restore(&mut self) -> Result<(), Error> {
+        info!(
+            "Starting snapshot restore from epoch {}",
+            self.restore_args.start_epoch
+        );
+        let (sha3_digests, num_part_files) = self.snapshot_reader.compute_checksum().await?;
+        let (_abort_handle, abort_registration) = AbortHandle::new_pair();
+        let (input_files, epoch_dir, remote_object_store, _concurrency) =
+            self.snapshot_reader.export_metadata().await?;
+        let owned_input_files: Vec<(u32, (u32, FileMetadata))> = input_files
+            .into_iter()
+            .map(|(bucket, (part_num, metadata))| (*bucket, (part_num, metadata.clone())))
+            .collect();
+
+        self.restore_object_infos(
+            abort_registration,
+            owned_input_files,
+            epoch_dir,
+            remote_object_store,
+            sha3_digests,
+            num_part_files,
+            self.restore_args.concurrency,
+        )
+        .await?;
+        info!(
+            "Finished snapshot restore from epoch {}",
+            self.restore_args.start_epoch
+        );
+        Ok(())
+    }
+
+    async fn restore_object_infos(
+        &self,
+        abort_registration: AbortRegistration,
+        input_files: Vec<(u32, (u32, FileMetadata))>,
+        epoch_dir: Path,
+        remote_object_store: Arc<dyn ObjectStoreGetExt>,
+        sha3_digests: Arc<Mutex<DigestByBucketAndPartition>>,
+        num_part_files: usize,
+        concurrency: usize,
+    ) -> std::result::Result<(), anyhow::Error> {
+        let move_object_progress_bar = Arc::new(self.snapshot_reader.get_multi_progress().add(
+            ProgressBar::new(num_part_files as u64).with_style(
+                ProgressStyle::with_template(
+                    "[{elapsed_precise}] {wide_bar} {pos} out of {len} move object files restored ({msg})",
+                )
+                .unwrap(),
+            ),
+        ));
+
+        Abortable::new(
+            async move {
+                let sema_limit = Arc::new(Semaphore::new(concurrency));
+                let mut restore_tasks = vec![];
+
+                for (bucket, (part_num, file_metadata)) in input_files.into_iter() {
```

Comment:
You could use `try_for_each_spawned` from the indexer framework to handle the concurrency limit, spawning and joining for you:

https://github.com/MystenLabs/sui/blob/8d4a73d3fb93690fbaddc0b75e49cc650b2e28dd/crates/sui-indexer-alt-framework/src/task.rs#L47-L51